master/internal/provisioner/aws_spot.go

```diff
@@ -0,0 +1,683 @@
+
+// The AWS API requires a validFrom time that is in the future according to AWS's clock.
+// See documentation of the spot struct for more detail. This function attempts
+// to create a spot request using the current values for c.spot.approximateClockSkew
+// and c.spot.launchTimeOffset. If that fails because AWS says the validFrom time is
+// not in the future, we increase c.spot.launchTimeOffset by launchTimeOffsetGrowth.
+// This can happen a maximum of 5 times before exiting with an error, to ensure that this
+// function doesn't block for too long.
+func (c *awsCluster) createSpotInstanceRequestsCorrectingForClockSkew(
+	ctx *actor.Context,
+	numInstances int,
+	dryRun bool,
+	instanceType ec2InstanceType,
+) (resp *ec2.RequestSpotInstancesOutput, err error) {
+	maxRetries := 5
+	for numRetries := 0; numRetries <= maxRetries; numRetries++ {
+		offset := c.spot.approximateClockSkew + c.spot.launchTimeOffset
+		resp, err = c.createSpotInstanceRequest(ctx, numInstances, instanceType, offset, dryRun)
+		if err == nil {
+			return resp, nil
+		}
+
+		if awsErr, ok := err.(awserr.Error); ok {
+			ctx.Log().Infof("AWS error while launching spot instances, %s, %s", awsErr.Code(), awsErr.Message())
+			if awsErr.Code() == "InvalidTime" {
+				c.spot.launchTimeOffset += launchTimeOffsetGrowth
+				ctx.Log().Infof("AWS error while launch spot instances - InvalidTime. Increasing "+
+					"launchOffset to %s to correct for clock skew",
+					c.spot.launchTimeOffset.String())
+			}
+		} else {
+			ctx.Log().Errorf("unknown error while launch spot instances, %s", err.Error())
+			return nil, err
+		}
+	}
+	return nil, err
+}
+
+func (c *awsCluster) createSpotInstanceRequest(
+	ctx *actor.Context,
+	numInstances int,
+	instanceType ec2InstanceType,
+	launchTimeOffset time.Duration,
+	dryRun bool,
+) (*ec2.RequestSpotInstancesOutput, error) {
+	if dryRun {
+		ctx.Log().Debug("dry run of createSpotInstanceRequest.")
+	}
+	idempotencyToken := uuid.New().String()
+
+	validFrom := time.Now().UTC().Add(c.spot.approximateClockSkew).Add(launchTimeOffset)
+	spotInput := &ec2.RequestSpotInstancesInput{
+		ClientToken:                  aws.String(idempotencyToken),
+		DryRun:                       aws.Bool(dryRun),
+		InstanceCount:                aws.Int64(int64(numInstances)),
+		InstanceInterruptionBehavior: aws.String("terminate"),
+		LaunchSpecification: &ec2.RequestSpotLaunchSpecification{
+			BlockDeviceMappings: []*ec2.BlockDeviceMapping{
+				{
+					DeviceName: aws.String("/dev/sda1"),
+					Ebs: &ec2.EbsBlockDevice{
+						DeleteOnTermination: aws.Bool(true),
+						VolumeSize:          aws.Int64(int64(c.RootVolumeSize)),
+						VolumeType:          aws.String("gp2"),
+					},
+				},
+			},
+			ImageId:      aws.String(c.ImageID),
+			InstanceType: aws.String(instanceType.name()),
+			KeyName:      aws.String(c.SSHKeyName),
+
+			UserData: aws.String(base64.StdEncoding.EncodeToString(c.ec2UserData)),
+		},
+		TagSpecifications: []*ec2.TagSpecification{
```

Comment:
we should also have tags for `Name`, the configured tag `tag_name`, and `determined-master-address` as on-demand instances.

Reply:
There are tags that are applied to the spot request, not the EC2 instance itself. We might want to apply tags to those EC2 instances themselves, but I believe that will require a bunch of API calls in `listSpot` and IMO that shouldn't block this PR (unless you disagree)

Reply:
I think it should be blocked since the instance names are configured via `instance_name` and it's fine to apply the same tags also to the spot requests since users also need a way to know which spot requests are managed by us.